config/routes.rb

```diff
@@ -17,6 +17,10 @@ def matches?(request)
   root to: redirect("/step-by-step-pages", status: 302)
 
   namespace :coronavirus do
+    resources :github_changes, only: %i[show update], param: :slug do
+      post "publish", to: "github_changes#publish", on: :member
+    end
```

Comment:
I feel like a resource might be the wrong way to define this compared to the approach we've taken with sub sections and timeline entries.

We'll end up with a URL here of `/coronavirus/github_changes/:slug` which feels quite different to `coronavirus/pages/:slug/sub_sections`.

I'd instead consider defining routes within the pages resource to be consistent with the other bits. I'd have thought you'd not really want a show route, that's perhaps more of an index (since there isn't necessarily a github_content resource to show) and have update and publish not need a parameter.

Reply:
Oh God I ummed and ahhed about this for ages. 

It's a weird one, because "github_changes" isn't part of the "page" model, which is what I assume the "pages" resource refers to, so it looked odd when I tried to add the resource under "pages" like announcements and timeline entries are etc.

I tried putting the routes under "pages" but that made the "update" and "publish" routes weird because they expect an id.

That's when I tried the current position of making it separate, but not quite like timeline entries.

Is your suggestion to define the individual routes inside "pages", rather than a resource? e.g.
```
resources :pages,
.
.
.
  get "/github_changes/:slug/"
  put "/github_changes/:slug/"
  post "/github_changes/:slug/"
```
etc?

However the route is defined, we will still need the slug.

Reply:
> It's a weird one, because "github_changes" isn't part of the "page" model, which is what I assume the "pages" resource refers to, so it looked odd when I tried to add the resource under "pages" like announcements and timeline entries are etc.

I get what you mean, I think announcements etc are all similarly part and not part of a page model too so it makes sense to handle them all consistently. With github_content we've got something that's not really very RESTful which always makes things harder.

By the way you get a slug by default by being within the `pages` resource.

I didn't have a particular suggestion for implementation but inline seems good. I guess that would come out with something like: 

```
    resources :pages,
              path: "",
              only: %i[index show update],
              param: :slug do
      get "github_content", to: "github_content#index", on: :member
      post "github_content/update", to: "github_content#update", as: "update_github_content", on: :member
      post "github_content/publish", to: "github_content#publish", as: "publish_github_content",  on: :member
```

producing:

```
➜  collections-publisher git:(timeline-summary) ✗ rails routes | grep github_content
              github_content_coronavirus_page GET    /coronavirus/:slug/github_content(.:format)                                          coronavirus/github_content#index
       update_github_content_coronavirus_page POST   /coronavirus/:slug/github_content/update(.:format)                                   coronavirus/github_content#update
      publish_github_content_coronavirus_page POST   /coronavirus/:slug/github_content/publish(.:format)                                  coronavirus/github_content#publish
```

And comparing that to a resources approach:

```
    resources :pages,
              path: "",
              only: %i[index show update],
              param: :slug do

      resources :github_content, only: %i[index] do
        post "update", as: :update, on: :collection
        post "publish", on: :collection
      end
```

You seem to get slightly odder route names out that have an `_index` suffix (this is [because `github_content` isn't a plural](https://github.com/rails/rails/blob/4e53e5ab0d15f3540d60bbbf4bcf70be575c5567/actionpack/lib/action_dispatch/routing/mapper.rb#L1218-L1220))

```
➜  collections-publisher git:(timeline-summary) ✗ rails routes | grep github_content
 update_coronavirus_page_github_content_index POST   /coronavirus/:page_slug/github_content/update(.:format)                              coronavirus/github_content#update
publish_coronavirus_page_github_content_index POST   /coronavirus/:page_slug/github_content/publish(.:format)                             coronavirus/github_content#publish
        coronavirus_page_github_content_index GET    /coronavirus/:page_slug/github_content(.:format)                                     coronavirus/github_content#index
```

The latter approach is probably more consistent with the rest of this app but is a bit more magical - and would require a bit of a fight to lose the `_index` suffix.

Reply:
Um, I called it `github_change` which is a plural.

Reply:
Oh great - that'd resolve the weird `_index` suffix then.

Reply:
Also, I think the convention is to use `show` if an id is being passed in: https://edgeguides.rubyonrails.org/routing.html#crud-verbs-and-actions

Reply:
I'm happy to give this a go though and see what happens:

```
resources :pages,
              path: "",
              only: %i[index show update],
              param: :slug do
      get "github_changes", to: "github_changes#show", on: :member
      post "github_changes/update", to: "github_changes#update", on: :member
      post "github_changes/publish", to: "github_changes#publish", on: :member
```

Reply:
The id isn't for the github_changes resource though, it's for the coronavirus page. Same way you can have /coronavirus/:slug/sub-sections which is an index.

Reply:
Great stuff 👍  (sorry my previous reply was out of sync your other one)

Reply:
I'm sure I tried using `index` in my first attempt and it wouldn't pass the slug through 🤔

Reply:
I've been having a little play about with the routes:

If they stay as they are, i.e.

```
namespace :coronavirus do
    resources :github_changes, only: %i[show update], param: :slug do
      post "publish", to: "github_changes#publish", on: :member
    end
```

We get:

```
publish_coronavirus_github_change  POST   /coronavirus/github_changes/:slug/publish(.:format)    coronavirus/github_changes#publish
coronavirus_github_change          GET    /coronavirus/github_changes/:slug(.:format)            coronavirus/github_changes#show
                                   PATCH  /coronavirus/github_changes/:slug(.:format)            coronavirus/github_changes#update
                                   PUT    /coronavirus/github_changes/:slug(.:format)            coronavirus/github_changes#update
```
Which gives us an extra unnecessary PATCH route.

Moving the individual routes inside the `pages` resource looks a bit messy in the routes file, but seems to give nicer routes:

```
namespace :coronavirus do    
    resources :pages,
              path: "",
              only: %i[index show],
              param: :slug do

      get "github_changes", to: "github_changes#show", on: :member
      put "github_changes/update", to: "github_changes#update", on: :member
      post "github_changes/publish", to: "github_changes#publish", on: :member


Gives:

github_changes_coronavirus_page GET           /coronavirus/:slug/github_changes(.:format)            coronavirus/github_changes#show
github_changes_update_coronavirus_page PUT    /coronavirus/:slug/github_changes/update(.:format)     coronavirus/github_changes#update
github_changes_publish_coronavirus_page POST  /coronavirus/:slug/github_changes/publish(.:format)    coronavirus/github_changes#publish

````

And `index` and `show` seem to be interchangeable when it comes to the path helper methods

namespace :coronavirus do    
    resources :pages,
              path: "",
              only: %i[index show],
              param: :slug do

      get "github_changes", to: "github_changes#index", on: :member
      put "github_changes/update", to: "github_changes#update", on: :member
      post "github_changes/publish", to: "github_changes#publish", on: :member


Gives:

github_changes_coronavirus_page         GET    /coronavirus/:slug/github_changes(.:format)           coronavirus/github_changes#index
github_changes_update_coronavirus_page  PUT    /coronavirus/:slug/github_changes/update(.:format)    coronavirus/github_changes#update
github_changes_publish_coronavirus_page POST   /coronavirus/:slug/github_changes/publish(.:format)   coronavirus/github_changes#publish

```

Adding the routes as a resource under `pages` gives some strange routes which require an id as well as the extra PATCH route:

```

namespace :coronavirus do    
    resources :pages,
              path: "",
              only: %i[index show],
              param: :slug do

      resources :github_changes, only: %i[index update] do
        post "publish", to: "github_changes#publish", on: :member
      end


Gives:

publish_coronavirus_page_github_change POST   /coronavirus/:page_slug/github_changes/:id/publish(.:format)   coronavirus/github_changes#publish
coronavirus_page_github_changes        GET    /coronavirus/:page_slug/github_changes(.:format)               coronavirus/github_changes#index coronavirus_page_github_change         PATCH  /coronavirus/:page_slug/github_changes/:id(.:format)           coronavirus/github_changes#update
                                       PUT    /coronavirus/:page_slug/github_changes/:id(.:format)                                 coronavirus/github_changes#update
```

If we add a param to the resource, which seems to defeat the purpose of nesting the resource in the first place, we no longer need to pass in an ID, but we still get the extra PATCH route:

```
namespace :coronavirus do    
    resources :pages,
              path: "",
              only: %i[index show],
              param: :slug do

      resources :github_changes, only: %i[index update], param: :slug do
        post "publish", to: "github_changes#publish", on: :member
      end


Gives:

publish_coronavirus_page_github_change POST   /coronavirus/:page_slug/github_changes/:slug/publish(.:format)   coronavirus/github_changes#publish
coronavirus_page_github_changes        GET    /coronavirus/:page_slug/github_changes(.:format)                 coronavirus/github_changes#index
coronavirus_page_github_change         PATCH  /coronavirus/:page_slug/github_changes/:slug(.:format)           coronavirus/github_changes#update
                                       PUT    /coronavirus/:page_slug/github_changes/:slug(.:format)           coronavirus/github_changes#update
```

In conclusion, my preference is add "individual" routes inside the `pages` resource.